src/package/packageAncestry.ts

```diff
@@ -0,0 +1,517 @@
+        MinorVersion: this.options?.node?.MinorVersion,
+        PatchVersion: this.options?.node?.PatchVersion,
+        BuildNumber: this.options?.node?.BuildNumber,
+        depthCounter: this.options?.depth,
+      },
+    };
+  }
+
+  public addNode(node: AncestryJsonProducer): void {
+    this.data.children.push(node.data);
+    this.children.push(node);
+  }
+
+  public search(version: string): AncestryJsonProducer {
+    const producers: AncestryJsonProducer[] = [];
+    producers.push(this);
+    while (producers.length > 0) {
+      const producer = producers.shift();
+      if (producer.label === version) {
+        return producer;
+      }
+      producers.push(...producer.children);
+    }
+    return undefined;
+  }
+
+  public produce<PackageAncestryNodeData>(): PackageAncestryNodeData {
+    return this.data.children[0] as unknown as PackageAncestryNodeData;
+  }
+}
+export class AncestryDotProducer implements AncestryRepresentationProducer {
+  public label: string;
+  public options: AncestryRepresentationProducerOptions;
+  private children: AncestryDotProducer[] = [];
+  public constructor(options?: AncestryRepresentationProducerOptions) {
+    this.options = options;
+    this.label = this.options?.node?.getVersion() || 'root';
+  }
+  /**
+   * Builds a node line in DOT, of the form nodeID [label="MAJOR.MINOR.PATCH"]
+   *
+   * @param currentNode
+   */
+  public static buildDotNode(currentNode: AncestryDotProducer): string {
+    return `\t node${currentNode.options.node.SubscriberPackageVersionId} [label="${currentNode.label}"]`;
+  }
+
+  /**
+   * Builds an edge line in DOT, of the form fromNode -- toNode
+   *
+   * @param fromNode
+   * @param toNode
+   */
+  public static buildDotEdge(fromNode: AncestryDotProducer, toNode: AncestryDotProducer): string {
+    return `\t node${fromNode.options.node.SubscriberPackageVersionId} -- node${toNode.options.node.SubscriberPackageVersionId}`;
+  }
+
+  public addNode(node: AncestryDotProducer): void {
+    this.children.push(node);
+  }
+
+  public produce(): string {
+    const producers: AncestryDotProducer[] = [];
+    producers.push(this);
+    const dotLines: string[] = [];
+
+    while (producers.length > 0) {
+      const producer = producers.shift();
+      if (producer.options) {
+        dotLines.push(AncestryDotProducer.buildDotNode(producer));
+      }
+      producers.push(...producer.children);
+    }
+    producers.push(this);
+    while (producers.length > 0) {
+      const producer = producers.shift();
+      if (producer.options) {
+        producer.children.forEach((child) => dotLines.push(AncestryDotProducer.buildDotEdge(producer, child)));
+      }
+      producers.push(...producer.children);
+    }
+    return `strict graph G {\n${dotLines.join('\n')}}`;
+  }
+}
+
+export class PackageAncestryNode extends AsyncCreatable<PackageAncestryNodeOptions> implements Attributes {
+  private readonly _version: VersionNumber;
```

Comment:
do we want to keep all of the underscores on private attributes?

Reply:
What do you suggest as an alternative?

Reply:
I haven't been prefixing private methods or attributes with anything, just letting the `private` scope take care of it. 🤷‍♂️ either way